torch/csrc/jit/type.h

```diff
@@ -287,42 +326,59 @@ struct TORCH_API NumberType : public Type {
   static const TypeKind Kind = TypeKind::NumberType;
   // global singleton
   static TypePtr get();
+private:
+  NumberType()
+  : Type(TypeKind::NumberType) {}
 };
 
+struct FloatType;
+using FloatTypePtr = std::shared_ptr<FloatType>;
 // This node represents a Python float number value
 struct TORCH_API FloatType : public Type {
-  FloatType()
-  : Type(TypeKind::FloatType) {}
+  template<typename ... T>
+  static FloatTypePtr create( T&& ... all ) {
+    return FloatTypePtr(new FloatType( std::forward<T>(all)... ));
+  }
   bool operator==(const Type& rhs) const override {
     return rhs.kind() == kind();
   }
   std::string str() const override {
     return "float";
   }
-  bool isSubtypeOf(const Type& rhs) const override {
-    return *this == rhs || rhs.kind() == TypeKind::NumberType;
+  bool isSubtypeOf(const TypePtr rhs) const override {
+    return *this == *rhs || rhs->kind() == TypeKind::NumberType;
   }
   static const TypeKind Kind = TypeKind::FloatType;
   // global singleton
   static TypePtr get();
+private:
+  FloatType()
+  : Type(TypeKind::FloatType) {}
 };
 
+struct IntType;
+using IntTypePtr = std::shared_ptr<IntType>;
 // This node represents a Python int number value
 struct TORCH_API IntType : public Type {
-  IntType()
-  : Type(TypeKind::IntType) {}
+  template<typename ... T>
+  static IntTypePtr create( T&& ... all ) {
+    return IntTypePtr(new IntType( std::forward<T>(all)... ));
+  }
   bool operator==(const Type& rhs) const override {
     return rhs.kind() == kind();
   }
   std::string str() const override {
     return "int";
   }
-  bool isSubtypeOf(const Type& rhs) const override {
-    return *this == rhs || rhs.kind() == TypeKind::NumberType;
+  bool isSubtypeOf(const TypePtr rhs) const override {
```

Comment:
Passing as `const TypePtr&` might be a bit faster because it doesn't need to increase and decrease the refcount. Also in other places where you're only reading it.